server/src/main/java/org/elasticsearch/index/mapper/IgnoredValuesFieldMapper.java

```diff
@@ -0,0 +1,132 @@
+    private static final int PARENT_OFFSET_IN_NAME_OFFSET = 1 << 16;
+
+    public static final String NAME = "_ignored_values";
+
+    public static final IgnoredValuesFieldMapper INSTANCE = new IgnoredValuesFieldMapper();
+
+    public static final TypeParser PARSER = new FixedTypeParser(context -> INSTANCE);
+
+    static final NodeFeature TRACK_IGNORED_VALUES = new NodeFeature("mapper.track_ignored_values");
+
+    /*
+     * Container for the ignored field data:
+     *  - the full name
+     *  - the offset in the full name indicating the end of the substring matching
+     *    the full name of the parent field
+     *  - the value, encoded as a byte array
+     */
+    public record NameValue(String name, int parentOffset, BytesRef value) {
+        String getParentFieldName() {
+            // _doc corresponds to the root object
+            return (parentOffset == 0) ? "_doc" : name.substring(0, parentOffset - 1);
+        }
+
+        String getFieldName() {
+            return parentOffset() == 0 ? name() : name().substring(parentOffset());
+        }
+    }
+
+    static final class IgnoredValuesFieldMapperType extends StringFieldType {
+
+        private static final IgnoredValuesFieldMapperType INSTANCE = new IgnoredValuesFieldMapperType();
+
+        private IgnoredValuesFieldMapperType() {
+            super(NAME, false, true, false, TextSearchInfo.NONE, Collections.emptyMap());
+        }
+
+        @Override
+        public String typeName() {
+            return NAME;
+        }
+
+        @Override
+        public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
+            return new StoredValueFetcher(context.lookup(), NAME);
+        }
+    }
+
+    private IgnoredValuesFieldMapper() {
+        super(IgnoredValuesFieldMapperType.INSTANCE);
+    }
+
+    @Override
+    protected String contentType() {
+        return NAME;
+    }
+
+    @Override
+    public void postParse(DocumentParserContext context) {
+        // Ignored values are only expected in synthetic mode.
+        assert context.getIgnoredFieldValues().isEmpty()
+            || context.indexSettings().getMode().isSyntheticSourceEnabled()
+            || context.mappingLookup().isSourceSynthetic();
+        for (NameValue nameValue : context.getIgnoredFieldValues()) {
+            context.doc().add(new StoredField(NAME, encode(nameValue)));
+        }
+    }
+
+    static byte[] encode(NameValue values) {
+        assert values.parentOffset < PARENT_OFFSET_IN_NAME_OFFSET;
+        assert values.parentOffset * (long) PARENT_OFFSET_IN_NAME_OFFSET < Integer.MAX_VALUE;
+
+        byte[] nameBytes = values.name.getBytes(StandardCharsets.UTF_8);
+        byte[] bytes = new byte[4 + nameBytes.length + values.value.length];
+        ByteUtils.writeIntLE(values.name.length() + PARENT_OFFSET_IN_NAME_OFFSET * values.parentOffset, bytes, 0);
+        System.arraycopy(nameBytes, 0, bytes, 4, nameBytes.length);
```

Comment:
It would be nice if we could reuse the name from `_ignored` doc values field (Salvatore's pr will store it as doc values instead of stored fields).  We end up storing it in _ignored too if number of fields is exceeded.

I think this is tricky, because then we don't know which value from `_ignored` field belongs to a value from`_ignored_values` field. In case of multi values for the same field, doc values store in alphabetic order.

Reply:
Yeah there's certain duplication here.. I added a note about this in the javadoc. I think this is good for now, let's get some mileage for this and optimize if we find out it's an issue in practice - ignored fields should be the exception after initial setup..

Reply:
Agreed, we can keep this in the back of our minds, and consider solutions to reduce storage for the two meta fields.